vmm/src/memory_manager.rs

```diff
@@ -1543,22 +1554,44 @@ impl Snapshottable for MemoryManager {
         let mut memory_manager_snapshot = Snapshot::new(MEMORY_MANAGER_SNAPSHOT_ID);
         let guest_memory = self.guest_memory.memory();
 
-        let mut memory_regions: Vec<MemoryRegion> = Vec::with_capacity(10);
+        let mut memory_regions: Vec<MemoryRegion> = Vec::new();
 
         guest_memory.with_regions_mut(|index, region| {
             if region.len() == 0 {
                 return Err(MigratableError::Snapshot(anyhow!("Zero length region")));
             }
 
+            let mut backing_file = Some(PathBuf::from(format!("memory-region-{}", index)));
+            if let Some(file_offset) = region.file_offset() {
+                if (region.flags() & libc::MAP_SHARED == libc::MAP_SHARED)
+                    && Self::file_backed_with_hardlink(file_offset.file())
```

Comment:
Why is it the case only when the file is backed by a hard link?

Reply:
When the file is backed by at least one hard link, that means the user has a way to retrieve access the file so that a restore will rely on it.
When the file is not backed by any hard link, that means CH created a temporary file with no link, either using the backing directory provided by the user or using `create_memfd()` function directly. In both these cases, there will be no resulting file that can be accessed by the user, which means these files won't be usable on the restore path, therefore we need to create a `memory-region-X` file for storing the content of this memory region.

Reply:
> When the file is backed by at least one hard link, that means the user has a way to retrieve access the file so that a restore will rely on it.

That makes sense now, thanks.